controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,14 +123,265 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO Filesystem volumes, and RWO and RWX Block volumes",
+			},
+			Finalizers: []string{
+				StorageClassFinalizer,
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.rbd.csi.ceph.com", r.storageCluster.Namespace),
+	}
+	for k, v := range data {
+		storageClass.Parameters[k] = v
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) createOrUpdateStorageClass(storageClass *storagev1.StorageClass) error {
+	existing := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: r.storageClassClaim.Name,
+		},
+	}
+	err := r.get(existing)
+	if errors.IsNotFound(err) {
+
+		// Since the StorageClass is not found, we will create a new one.
+		r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+		err = r.Client.Create(r.ctx, storageClass)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	} else {
+		if existing.DeletionTimestamp != nil {
+			return fmt.Errorf("failed to restore StorageClass  %s because it is marked for deletion", existing.Name)
+		}
+		if !reflect.DeepEqual(storageClass.Parameters, existing.Parameters) {
+
+			// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
+			r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+
+			// Check if StorageClass is in use, if yes, then fail.
+			// Wait until all PVs using the StorageClass under deletion are removed.
+			// Check for any PVs using the StorageClass.
+			pvList := &corev1.PersistentVolumeList{}
+			err := r.list(pvList, client.MatchingFields{
+				"spec.storageClassName": r.storageClassClaim.Name,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+			}
+			if len(pvList.Items) > 0 {
+				return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+			}
+
+			// Delete the StorageClass.
+			err = r.delete(existing)
+			if err != nil {
+				r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+				return err
+			}
+			r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, storageClass.Name))
+			err = r.Client.Create(r.ctx, storageClass)
+			if err != nil {
+				r.log.Info("Failed to create StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, storageClass.Name))
+				return err
+			}
+		}
+	}
+	return nil
```

Comment:
I find the here confusing and hard to follow, it has at least two separate create sections that could be merged. 
I would suggest refactoring according to the following flow:

```
	// Fetch existing claim if any
	existing := &storagev1.StorageClass{}
	existing.Name = r.storageClassClaim.Name
	if err := r.get(existing); err != nil && !errors.IsNotFound(err) {
		return err
	}

	// Already exists, need to delete it first.
	if existing.UID != "" {
		// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
		r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))

		// Check if StorageClass is in use, if yes, then fail.
		// Wait until all PVs using the StorageClass under deletion are removed.
		// Check for any PVs using the StorageClass.
		pvList := corev1.PersistentVolumeList{}
		err := r.list(&pvList, client.MatchingFields{
			"spec.storageClassName": r.storageClassClaim.Name,
		})
		if err != nil {
			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
		}
		if len(pvList.Items) > 0 {
			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
		}

		// Delete the StorageClass.
		err = r.delete(existing)
		if err != nil {
			r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
			return err
		}
	}

	// Create the new storage class claim
	r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
	err = r.Client.Create(r.ctx, storageClass)
	if err != nil {
		return err
	}

	return nil
```

But even in this case you need to handle the finalizer on the storageclass itself.